am2r_yams/Patcher.py

```diff
@@ -9,69 +9,80 @@
 from typing import Callable
 
 # TODO: revise this to be cleaner. Following things to keep in mind here
-# 1. AFAIK the DLLs must be in path before loading the CLR. test whether that's truly the case
-# 2. the CLR should only loaded *once*. I.e importing this module multiple times shouldnt cause exceptions of trying to load it multiple times.
-# 3. Figure out how to deal with non-system dotnet installations
+# 1. AFAIK the DLLs must be in path before loading the CLR. test whether that's truly the case.
+#    Maybe we can remove from path when we're done?
+# 2. Figure out how to deal with non-system dotnet installations
 yams_path = os.fspath(Path(__file__).with_name(name="yams"))
 sys.path.append(yams_path)
-from pythonnet import load
-
-load("coreclr")
-import clr
-
-clr.AddReference("YAMS-LIB")
-from YAMS_LIB import Patcher
-
-
-def get_lib_version() -> str:
-    return Patcher.Version
-
-
-# TODO: ADD TESTS!!!
-def patch_game(
-    input_path: Path,
-    output_path: Path,
-    patch_data: dict,
-    progress_update: Callable[[str, float], None],
-):
-    # Copy to input dir to temp dir first to do operations there
-    progress_update("Copying to temporary path...", 0)
-    tempdir = tempfile.TemporaryDirectory()
-    shutil.copytree(input_path, tempdir.name, dirs_exist_ok=True)
-
-    # Get data.win path. Both of these *need* to be strings, as otherwise patcher won't accept them.
-    output_data_win: str = os.fspath(
-        _prepare_environment_and_get_data_win_path(tempdir.name)
-    )
-    input_data_win: str = shutil.move(output_data_win, output_data_win + "_orig")
-    input_data_win_path = Path(input_data_win)
-
-    # TODO: do some check on whether input is valid? on patcher side probably.
-
-    # Temp write patch_data into json file for yams later
-    progress_update("Creating json file...", 0.3)
-    json_file: str = os.fspath(input_data_win_path.parent.joinpath("yams-data.json"))
-    with open(json_file, "w+") as f:
-        f.write(json.dumps(patch_data))
-
-    # AM2RLauncher installations usually have a profile.xml file. For less confusion, remove it if it exists
-    profile_xml_path = Path(tempdir.name).joinpath("profile.xml")
-    if profile_xml_path.exists():
-        profile_xml_path.unlink()
-
-    # TODO: this is where we'd call the patcher to do some customization options like music shuffler or samus palettes
-
-    # Patch data.win
-    progress_update("Patching data file...", 0.6)
-    Patcher.Main(input_data_win, output_data_win, json_file)
-
-    # Move temp dir to output dir and get rid of it. Also delete original data.win
-    input_data_win_path.unlink()
-    progress_update("Moving to output directory...", 0.8)
-    shutil.copytree(tempdir.name, output_path, dirs_exist_ok=True)
-    shutil.rmtree(tempdir.name)
-
-    progress_update("Exporting finished!", 1)
+from pythonnet import load, unload
+
+
+# TODO: add docstrings for methods when not in alpha and has stableish API
+class Patcher:
```

Comment:
I would make this class just take a "yasm_lib" as argument to the constructor, which it then stores in a field.
`patch_game` would then use that to find the main method.

All the loading/unloading of the CLR would be done by an additional method, which you'd use `contextlib.contextmanager`

Reply:
Just so i get this correctly, something like:
```python

@contextmanager
def load_lib():
  try:
    # load runtime
    from YAMS_Lib import Patcher
    yield Patcher
  finally:
    # unload runtime

class Patcher():
  def __init__(self, lib):
    self.patcher_lib = lib

  def get_lib_version():
    return self.patcher_lib.Version
```
With usage being
```python
with load_lib() as lib:
    patcher = Patcher(lib)
    version = patcher.get_lib_version
```

?

Reply:
usage would be just:
```python
with load_lib() as patcher:
    version = patcher.get_lib_version()
```